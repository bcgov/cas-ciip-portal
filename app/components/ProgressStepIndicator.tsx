import React from 'react';
import {Col, ProgressBar, Row} from 'react-bootstrap';
import {Variant} from 'react-bootstrap/esm/types';

interface NumberCircleProps {
  number: number;
  style: Variant;
}

interface Props {
  steps: Array<{
    description: string;
    badgeStyle: Variant;
  }>;
}

const NumberedCircle: React.FunctionComponent<NumberCircleProps> = ({
  number,
  style
}) => {
  return (
    <>
      <div className={`numberedCircle badge-${style}`}>{number}</div>
      <style jsx>{`
        .numberedCircle {
          margin-top: -29px;
          width: 38px;
          line-height: 37px;
          border-radius: 50%;
          text-align: center;
          font-size: 16px;
          font-weight: bold;
        }
      `}</style>
    </>
  );
};

export const ProgressStepIndicator: React.FunctionComponent<Props> = ({
  steps
}) => {
  return (
    <>
      <div>
        <ProgressBar now={0} />
        <div className="d-flex flex-row justify-content-between mb-3">
          {steps.map((step, index) => (
            <NumberedCircle number={index + 1} style={step.badgeStyle} />
          ))}
        </div>
        <Row>
          {steps.map((step) => (
            <Col className="stepDescription">{step.description}</Col>
          ))}
        </Row>
      </div>
      <style jsx global>{`
        .stepDescription {
          text-align: center;
          font-weight: 500;
        }
        .stepDescription:first-child {
          text-align: left;
        }
        .stepDescription:last-child {
          text-align: right;
        }
      `}</style>
    </>
  );
};

export default ProgressStepIndicator;
