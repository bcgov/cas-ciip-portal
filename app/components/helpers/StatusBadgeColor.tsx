export default {
  INITIAL: 'warning',
  REJECTED: 'danger',
  PENDING: 'info',
  APPROVED: 'success'
};
