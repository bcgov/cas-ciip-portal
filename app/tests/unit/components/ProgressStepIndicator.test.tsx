import StatusBadgeColor from 'components/helpers/StatusBadgeColor';
import ProgressStepIndicator from 'components/ProgressStepIndicator';
import {mount} from 'enzyme';
import React from 'react';

describe('The Progress Step Indicator', () => {
  it('Renders with no steps', () => {
    const componentUnderTest = mount(<ProgressStepIndicator steps={[]} />);
    expect(componentUnderTest).toMatchSnapshot();
  });

  it('Matches snapshot with 3 steps and a title', () => {
    const componentUnderTest = mount(
      <ProgressStepIndicator
        title="This is a TEST TITLE"
        steps={[
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 1',
            number: 99
          },
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 2',
            number: 100
          },
          {
            badgeStyle: StatusBadgeColor.REJECTED,
            description: 'test step 3',
            number: 110
          }
        ]}
      />
    );
    expect(componentUnderTest).toMatchSnapshot();
    expect(componentUnderTest.find('.numberedCircle')).toHaveLength(3);
    expect(componentUnderTest.find('div.stepDescription')).toHaveLength(3);
    const progressBar = componentUnderTest.find('.progress-bar');
    expect(progressBar.getDOMNode().getAttribute('aria-valuenow')).toEqual('0');
    expect(progressBar.getDOMNode().getAttribute('aria-valuetext')).toEqual(
      'test step 1'
    );
    expect(progressBar.prop('style')).toHaveProperty('width', '0%');
  });

  const stepCountTestCases = [
    [1, StatusBadgeColor.APPROVED],
    [2, StatusBadgeColor.PENDING],
    [3, StatusBadgeColor.INITIAL],
    [4, StatusBadgeColor.REJECTED],
    [5, StatusBadgeColor.APPROVED],
    [6, StatusBadgeColor.PENDING]
  ];
  test.each(stepCountTestCases)(
    'Generates %p progress items',
    (stepCount: number, status: string) => {
      const steps = Array.from({length: stepCount}, (_, index) => ({
        badgeStyle: status,
        description: `test step ${index}`,
        number: index
      }));

      const componentUnderTest = mount(<ProgressStepIndicator steps={steps} />);
      expect(componentUnderTest.find('.numberedCircle')).toHaveLength(
        stepCount
      );
      componentUnderTest.find('.numberedCircle').forEach((wrapper, index) => {
        expect(wrapper.text()).toBe(`${index}`);
        expect(wrapper.hasClass('badge-' + status)).toBeTrue();
      });

      expect(componentUnderTest.find('div.stepDescription')).toHaveLength(
        stepCount
      );
      componentUnderTest
        .find('div.stepDescription')
        .forEach((wrapper, index) => {
          expect(wrapper.text()).toBe('test step ' + index);
        });
    }
  );

  it('Matches snapshot with 2/4 steps completed and a title', () => {
    const componentUnderTest = mount(
      <ProgressStepIndicator
        title="This is a TEST TITLE"
        steps={[
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 1',
            number: 99,
            completed: true
          },
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 2',
            number: 100,
            completed: true
          },
          {
            badgeStyle: StatusBadgeColor.REJECTED,
            description: 'test step 3',
            number: 110
          },
          {
            badgeStyle: StatusBadgeColor.REJECTED,
            description: 'test step 3',
            number: 111
          }
        ]}
      />
    );
    expect(componentUnderTest).toMatchSnapshot();
    expect(componentUnderTest.find('.numberedCircle')).toHaveLength(4);
    expect(componentUnderTest.find('div.stepDescription')).toHaveLength(4);
    expect(componentUnderTest.find('.fa-check')).toHaveLength(2);

    const progressBar = componentUnderTest.find('.progress-bar');
    expect(progressBar.getDOMNode().getAttribute('aria-valuenow')).toEqual('1');
    expect(progressBar.getDOMNode().getAttribute('aria-valuetext')).toEqual(
      'test step 2'
    );
    expect(progressBar.prop('style')).toHaveProperty('width', '33%');
  });

  it('Matches snapshot with 4/4 steps completed and a title', () => {
    const componentUnderTest = mount(
      <ProgressStepIndicator
        title="This is a TEST TITLE"
        steps={[
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 1',
            number: 99,
            completed: true
          },
          {
            badgeStyle: StatusBadgeColor.APPROVED,
            description: 'test step 2',
            number: 100,
            completed: true
          },
          {
            badgeStyle: StatusBadgeColor.REJECTED,
            description: 'test step 3',
            number: 110,
            completed: true
          },
          {
            badgeStyle: StatusBadgeColor.REJECTED,
            description: 'test step 4',
            number: 111,
            completed: true
          }
        ]}
      />
    );
    expect(componentUnderTest).toMatchSnapshot();
    expect(componentUnderTest.find('.numberedCircle')).toHaveLength(4);
    expect(componentUnderTest.find('div.stepDescription')).toHaveLength(4);
    expect(componentUnderTest.find('.fa-check')).toHaveLength(4);

    const progressBar = componentUnderTest.find('.progress-bar');
    expect(progressBar.getDOMNode().getAttribute('aria-valuenow')).toEqual('3');
    expect(progressBar.getDOMNode().getAttribute('aria-valuetext')).toEqual(
      'test step 4'
    );
    expect(progressBar.prop('style')).toHaveProperty('width', '100%');
  });
});
