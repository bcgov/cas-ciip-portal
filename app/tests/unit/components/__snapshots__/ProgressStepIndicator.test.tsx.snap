// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Progress Step Indicator Matches snapshot with 2/4 steps completed and a title 1`] = `
<div
  className="jsx-1128701548 progressStepIndicator"
>
  <Row
    className="mb-4"
    noGutters={false}
  >
    <div
      className="mb-4 row"
    >
      <Col>
        <div
          className="col"
        >
          <h2
            className="jsx-1128701548"
          >
            This is a TEST TITLE
          </h2>
        </div>
      </Col>
    </div>
  </Row>
  <div
    className="jsx-1128701548 progress ml-1 mr-1"
  >
    <div
      aria-label="progress bar indicating the steps represented by this component"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={33}
      className="jsx-1128701548 progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "33%",
        }
      }
    />
  </div>
  <div
    className="jsx-1128701548 d-flex flex-row justify-content-between mb-3"
  >
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-danger"
      >
        110
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-danger"
      >
        111
      </div>,
      "",
    ]
  </div>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Col
        className="stepDescription"
        key="description-99"
      >
        <div
          className="stepDescription col"
        >
          test step 1
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-100"
      >
        <div
          className="stepDescription col"
        >
          test step 2
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-110"
      >
        <div
          className="stepDescription col"
        >
          test step 3
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-111"
      >
        <div
          className="stepDescription col"
        >
          test step 3
        </div>
      </Col>
    </div>
  </Row>
</div>
`;

exports[`The Progress Step Indicator Matches snapshot with 3 steps and a title 1`] = `
<div
  className="jsx-1128701548 progressStepIndicator"
>
  <Row
    className="mb-4"
    noGutters={false}
  >
    <div
      className="mb-4 row"
    >
      <Col>
        <div
          className="col"
        >
          <h2
            className="jsx-1128701548"
          >
            This is a TEST TITLE
          </h2>
        </div>
      </Col>
    </div>
  </Row>
  <div
    className="jsx-1128701548 progress ml-1 mr-1"
  >
    <div
      aria-label="progress bar indicating the steps represented by this component"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="jsx-1128701548 progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "0%",
        }
      }
    />
  </div>
  <div
    className="jsx-1128701548 d-flex flex-row justify-content-between mb-3"
  >
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        99
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        100
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-danger"
      >
        110
      </div>,
      "",
    ]
  </div>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Col
        className="stepDescription"
        key="description-99"
      >
        <div
          className="stepDescription col"
        >
          test step 1
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-100"
      >
        <div
          className="stepDescription col"
        >
          test step 2
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-110"
      >
        <div
          className="stepDescription col"
        >
          test step 3
        </div>
      </Col>
    </div>
  </Row>
</div>
`;

exports[`The Progress Step Indicator Matches snapshot with 4/4 steps completed and a title 1`] = `
<div
  className="jsx-1128701548 progressStepIndicator"
>
  <Row
    className="mb-4"
    noGutters={false}
  >
    <div
      className="mb-4 row"
    >
      <Col>
        <div
          className="col"
        >
          <h2
            className="jsx-1128701548"
          >
            This is a TEST TITLE
          </h2>
        </div>
      </Col>
    </div>
  </Row>
  <div
    className="jsx-1128701548 progress ml-1 mr-1"
  >
    <div
      aria-label="progress bar indicating the steps represented by this component"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={100}
      className="jsx-1128701548 progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div
    className="jsx-1128701548 d-flex flex-row justify-content-between mb-3"
  >
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-success"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-danger"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
    Array [
      <div
        className="jsx-1375059393 numberedCircle badge-danger"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "fontSize": "30px",
              "height": "0.8em",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>,
      "",
    ]
  </div>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Col
        className="stepDescription"
        key="description-99"
      >
        <div
          className="stepDescription col"
        >
          test step 1
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-100"
      >
        <div
          className="stepDescription col"
        >
          test step 2
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-110"
      >
        <div
          className="stepDescription col"
        >
          test step 3
        </div>
      </Col>
      <Col
        className="stepDescription"
        key="description-111"
      >
        <div
          className="stepDescription col"
        >
          test step 3
        </div>
      </Col>
    </div>
  </Row>
</div>
`;

exports[`The Progress Step Indicator Renders with no steps 1`] = `
<div
  className="jsx-1128701548 progressStepIndicator"
>
  <div
    className="jsx-1128701548 progress ml-1 mr-1"
  >
    <div
      aria-label="progress bar indicating the steps represented by this component"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="jsx-1128701548 progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "0%",
        }
      }
    />
  </div>
  <div
    className="jsx-1128701548 d-flex flex-row justify-content-between mb-3"
  />
  <Row
    noGutters={false}
  >
    <div
      className="row"
    />
  </Row>
</div>
`;
